src/Tools/dotnet-monitor/Egress/EgressService.cs

```diff
@@ -16,31 +21,129 @@ namespace Microsoft.Diagnostics.Tools.Monitor.Egress
     /// <summary>
     /// Egress service implementation required by the REST server.
     /// </summary>
-    internal class EgressService : IEgressService
+    internal class EgressService : IEgressService, IDisposable
     {
-        private readonly IOptionsMonitor<EgressOptions> _egressOptions;
+        private readonly IEnumerable<IEgressProviderConfigurationProvider> _providers;
+        private readonly IDictionary<string, string> _providerTypeMap;
+        private readonly IDictionary<string, Type> _optionsTypeMap;
+        private readonly IList<IDisposable> _registrations;
+        private readonly IServiceProvider _serviceProvider;
 
-        public EgressService(IOptionsMonitor<EgressOptions> egressOptions)
+        public EgressService(
+            IServiceProvider serviceProvider,
+            IEnumerable<IEgressProviderConfigurationProvider> providers)
         {
-            _egressOptions = egressOptions;
+            _providers = providers;
+            _providerTypeMap = new ConcurrentDictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+            _optionsTypeMap = new ConcurrentDictionary<string, Type>(StringComparer.OrdinalIgnoreCase);
+            _registrations = new List<IDisposable>();
+            _serviceProvider = serviceProvider;
+
+            foreach (IEgressProviderConfigurationProvider provider in providers)
+            {
+                _registrations.Add(ChangeToken.OnChange(
+                    () => provider.Configuration.GetReloadToken(),
+                    () => Reload()));
+            }
+
+            Reload();
+        }
+
+        public void Dispose()
+        {
+            foreach (IDisposable registration in _registrations)
+            {
+                registration.Dispose();
+            }
+            _registrations.Clear();
+        }
+
+        public async Task<EgressResult> EgressAsync(string providerName, Func<CancellationToken, Task<Stream>> action, string fileName, string contentType, IEndpointInfo source, CancellationToken token)
+        {
+            string value = await GetProvider(providerName).EgressAsync(
+                providerName,
+                action,
+                CreateSettings(source, fileName, contentType),
+                token);
+
+            return new EgressResult("name", value);
+        }
+
+        public async Task<EgressResult> EgressAsync(string providerName, Func<Stream, CancellationToken, Task> action, string fileName, string contentType, IEndpointInfo source, CancellationToken token)
+        {
+            string value = await GetProvider(providerName).EgressAsync(
+                providerName,
+                action,
+                CreateSettings(source, fileName, contentType),
+                token);
+
+            return new EgressResult("name", value);
```

Comment:
I know you left feedback on my PR to convert this to just a string indicating the location, but I would like to leave it as an object in case we end up putting more information on it. The object will be uniform between egress types (at least for now)